vortex-array/src/array/varbinview/builder.rs

```diff
@@ -0,0 +1,111 @@
+use std::marker::PhantomData;
+use std::mem;
+use std::mem::ManuallyDrop;
+
+use arrow_buffer::NullBufferBuilder;
+use vortex_schema::DType;
+
+use crate::array::primitive::PrimitiveArray;
+use crate::array::varbinview::{BinaryView, Inlined, Ref, VarBinViewArray, VIEW_SIZE};
+use crate::array::{Array, ArrayRef, IntoArray};
+use crate::validity::Validity;
+
+pub struct VarBinViewBuilder<T: AsRef<[u8]>> {
+    views: Vec<BinaryView>,
+    nulls: NullBufferBuilder,
+    completed: Vec<ArrayRef>,
+    in_progress: Vec<u8>,
+    block_size: u32,
+    phantom: PhantomData<T>,
+}
+
+impl<T: AsRef<[u8]>> VarBinViewBuilder<T> {
+    pub fn with_capacity(capacity: usize) -> Self {
+        Self {
+            views: Vec::with_capacity(capacity),
+            nulls: NullBufferBuilder::new(capacity),
+            completed: Vec::new(),
+            in_progress: Vec::new(),
+            block_size: 16 * 1024,
+            phantom: Default::default(),
+        }
+    }
+
+    #[inline]
+    pub fn push(&mut self, value: Option<T>) {
+        match value {
+            None => self.push_null(),
+            Some(v) => self.push_value(v),
+        }
+    }
+
+    #[inline]
+    pub fn push_value(&mut self, value: T) {
+        let vbytes = value.as_ref();
+        if self.in_progress.len() + vbytes.len() > self.in_progress.capacity() {
+            let done = mem::replace(
+                &mut self.in_progress,
+                Vec::with_capacity(vbytes.len().max(self.block_size as usize)),
+            );
+            if !done.is_empty() {
+                assert!(self.completed.len() < u32::MAX as usize);
+                self.completed.push(PrimitiveArray::from(done).into_array());
+            }
+        }
+
+        if vbytes.len() > BinaryView::MAX_INLINED_SIZE {
+            self.views.push(BinaryView {
+                _ref: Ref::new(
+                    vbytes.len() as u32,
+                    vbytes[0..4].try_into().unwrap(),
```

Comment:
magic number 4 should probably be a named constant in the BinaryView namespace, no?